### PR TITLE
Error on yarn install if the lockfile is out of date

### PR DESCRIPTION
**Pull request by jmorrell**
This switches to using `--frozen-lockfile` for yarn installs, which will fail the build if `yarn.lock` does not match the requirements listed in `package.json`. Under our current setup Yarn will install the updated dependencies as if the lockfile doesn't exist in this case, which could lead to unexpected failures at runtime.

With this change it will fail the build and print out a helpful message in this case.

Example log and error message:
```
-----> Creating runtime environment
       
       NPM_CONFIG_LOGLEVEL=error
       NPM_CONFIG_PRODUCTION=true
       NODE_VERBOSE=false
       NODE_ENV=production
       NODE_MODULES_CACHE=true

-----> Installing binaries
       engines.node (package.json):  unspecified
       engines.npm (package.json):   unspecified (use default)
       engines.yarn (package.json):  unspecified (use default)
       
       Resolving node version 6.x via semver.io...
       Downloading and installing node 6.11.1...
       Using default npm version: 3.10.10
       Resolving yarn version (latest)...
       Downloading and installing yarn (0.28.1)...
       Installed yarn 0.28.1

-----> Restoring cache
       Skipping cache restore (not-found)

-----> Building dependencies
       Installing node modules (yarn.lock)
       yarn install v0.28.1
       [1/4] Resolving packages...
       error Your lockfile needs to be updated, but yarn was run with `--frozen-lockfile`.
       info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

-----> Build failed

 !     Outdated Yarn lockfile

       Your application contains a Yarn lockfile: yarn.lock which does not match
       the requirements in your package.json file. This can happen if you use
       npm to install or update a dependency instead of Yarn.

       Please run the following command in your application directory and check
       in the new yarn.lock file:

       $ yarn install
       $ git add yarn.lock
       $ git commit -m "Updated Yarn lockfile"
       $ git push heroku master
    
       https://kb.heroku.com/why-is-my-node-js-build-failing-because-of-an-outdated-yarn-lockfile
```

### CODE REVIEW

**Comment by jmorrell** [comment]
Tests are currently failing because yarn fails on unknown command line flags, so I need to avoid passing in `--frozen-lockfile` for yarn versions older than X. Will fix.

File: lib/dependencies.sh [hunk 1]
```diff
@@ -59,12 +59,28 @@ log_build_scripts() {
   fi
 }
 
+yarn_supports_frozen_lockfile() {
```
Comment (jmorrell) [inline lib/dependencies.sh]: @dzuelke Mind critiquing my bash here?

File: lib/dependencies.sh [hunk 2]
```diff
@@ -59,12 +59,28 @@ log_build_scripts() {
...
+  local yarn_version="$(yarn --version)"
+  # Yarn versions lower than 0.19 will crash if passed --frozen-lockfile
+  if [[ "$yarn_version" =~ ^0\.(16|17|18).*$ ]]; then
```
Comment (jmorrell) [inline lib/dependencies.sh]: List of all the yarn versions we support: https://nodebin.herokai.com/v1/yarn/linux-x64.txt

File: lib/dependencies.sh [hunk 3]
```diff
@@ -59,12 +59,28 @@ log_build_scripts() {
...
+    mcount "yarn.doesnt-support-frozen-lockfile"
+    echo false
```
Comment (dzuelke) [inline lib/dependencies.sh]: Use `return 1`, or just `false`, without `echo`

File: lib/dependencies.sh [hunk 4]
```diff
@@ -59,12 +59,28 @@ log_build_scripts() {
...
+  else
+    echo true
```
Comment (dzuelke) [inline lib/dependencies.sh]: Use `return 0`, or just `true`, without `echo`

File: lib/dependencies.sh [hunk 5]
```diff
@@ -59,12 +59,28 @@ log_build_scripts() {
...
+}
+
 yarn_node_modules() {
   local build_dir=${1:-}
+  local supports_frozen_lockfile=$(yarn_supports_frozen_lockfile)
 
   echo "Installing node modules (yarn.lock)"
   cd "$build_dir"
-  yarn install --pure-lockfile --ignore-engines 2>&1
+  if [ "$supports_frozen_lockfile" = true ]; then
```
Comment (dzuelke) [inline lib/dependencies.sh]: `if yarn_supports_frozen_lockfile; then` (and, generally, always use `[[` with `if` if you require statements inside)

File: lib/dependencies.sh [hunk 6]
```diff
@@ -59,12 +59,28 @@ log_build_scripts() {
...
+    mcount "yarn.doesnt-support-frozen-lockfile"
+    echo false
+  else
+    echo true
+  fi
+}
+
 yarn_node_modules() {
   local build_dir=${1:-}
+  local supports_frozen_lockfile=$(yarn_supports_frozen_lockfile)
```
Comment (dzuelke) [inline lib/dependencies.sh]: You don't need that one; see below

**Comment by jmorrell** [comment]
> If you didn't need the mcount … call, the last statement could thus also just be [[ ! "$yarn_version" =~ ^0\.(16|17|18).*$ ]]

I see how I could simplify to that, but I like the clarity of the more verbose option. Thanks for the insights @dzuelke!

File: lib/failure.sh [hunk 7]
```diff
@@ -116,6 +116,29 @@ fail_multiple_lockfiles() {
...
 }
 
+fail_yarn_lockfile_outdated() {
+  local log_file="$1"
+  if grep -qi 'error Your lockfile needs to be updated' "$log_file"; then
+    mcount "failures.outdated-yarn-lockfile"
+    echo ""
+    warn "Outdated Yarn lockfile
+
+       Your application contains a Yarn lockfile: yarn.lock which does not match
```
Comment (hunterloftis) [inline lib/failure.sh]: Parentheses might fit the grammar here better than a colon:

> Your application contains a Yarn lockfile; yarn.lock which does not match the requirements in your package.json file.

vs:

> Your application contains a lockfile (yarn.lock) which does not match the dependencies in package.json.

**Comment by jmorrell** [comment]
updated wording 👍  @hunterloftis

**Comment by hone** [comment]
Thanks for getting this change in! sorry for not reviewing it sooner.

**Comment by vlad-elagin** [comment]
Hello.
Can't get my app working with error listed by topic-starter. I tried to delete and re-create lockfile, but without success. Can you help me with this? If so, what information should I provide?

**Comment by hone** [comment]
@vlad-elagin hi, can you provide your `package.json` and `yarn.lock`?

**Comment by vlad-elagin** [comment]
Thanks for response! Here they are.

[yarn.zip](https://github.com/heroku/heroku-buildpack-nodejs/files/1205707/yarn.zip)

**Comment by vlad-elagin** [comment]
I've got it working, thanks.

**Pull request merged**